core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -483,14 +493,10 @@ class RequestChannel(val queueSize: Int,
 
   def sendShutdownRequest(): Unit = requestQueue.put(ShutdownRequest)
 
-  // TODO: this is the most straightforward implementation, we may want to address fairness:
-  // The callback function is invoked as part of processing a request, so it already "stood in line"
-  // when the original request got accepted, so it may be unfair to make put to the end of the line.
-  // A simple solution would be to put it to the head of the line, but then multiple callbacks would
-  // be in LIFO order, which is weird.
-  // We may want to use a PriorityBlockingQueue or just have 2 queues and write customer wait / notify
-  // synchronization.
-  def sendCallbackRequest(request: CallbackRequest): Unit = requestQueue.put(request)
+  def sendCallbackRequest(request: CallbackRequest): Unit = {
+    callbackQueue.put(request)
```

Comment:
Could we just reuse the ActionQueue instead of introducing a new queue? ActionQueue is drained by the request thread whenever it finishes processing the current event.

Reply:
Will we still have the behavior of putting the callback at the front of the queue? It would go to the end of the action queue right?

Reply:
It would go to the end of the action queue. However, not every request adds entries to the action queue. So, the action queue is typically smaller than the request queue. Also, request threads prioritize the action queue over the request queue.

Reply:
thanks!

Reply:
I'm going to keep what I have for now -- we can revisit the action queue if we want later.